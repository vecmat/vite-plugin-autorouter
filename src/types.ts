export type ImportMode = "sync" | "async";
export type ImportModeResolveFn = (filepath: string) => ImportMode;


export interface ResolvedPage extends Record<string, any> {
    name?: string;
    path: string;
    component: string;
}
export type ResolvedPages = Map<string, ResolvedPage>;


export interface Route {
    // custom
    chain?: string;
    paths?: string[];

    // basic
    path: string;
    name?: string;
    alias?: string|string[];
    props?: boolean | Record<string, any> | ((to: any) => Record<string, any>);
    component: string;
    exact?: boolean;
    routes?: Route[];
    beforeEnter?: any;
    children?: Route[];
    meta?: Record<string, unknown>;
}


export interface ResolvedOptions extends Options {
    root: string;
    pagesDir: string[];
    extensionsRE: RegExp;
}


/**
 * Plugin options.
 */
 interface Options {
    /**
     * Relative path to the directory to search for page components.
     * @default 'src/pages'
     */
    pagesDir: string[];
    /**
     * Valid file extensions for page components.
     * @default ['vue', 'js']
     */
    extensions: string[];
    /**
     * List of path globs to exclude when resolving pages.
     */
    exclude: string[];
    /**
     * Import routes directly or as async components
     * @default 'async'
     */
    importMode: ImportMode | ImportModeResolveFn;
    /**
     * Sync load top level index file
     * @default true
     */
    syncIndex: boolean;
    /**
     * Use Nuxt.js style dynamic routing
     * @default false
     */
    nuxtStyle: boolean;
    /**
     * Set default route block parser, or use `<route lang=xxx>` in SFC route block
     * @default 'json5'
     */
    routeBlockLang: "json5" | "json" | "yaml" | "yml" | "toml";
    /**
     * Replace '[]' to '_' in bundle chunk filename
     * Experimental feature
     * @default true
     */
    replaceSquareBrackets: boolean;
    /**
     * Generate React Route
     * @default false
     */
    react: boolean;
    /**
     * Extend route records
     */
    extendRoute?: (route: Route) => Route | void;
    /**
     * Custom generated routes
     */
    onRoutesGenerated?: (routes: Route[]) => Route[] | void | Promise<Route[] | void>;
    /**
     * Custom generated client code
     */
    onClientGenerated?: (clientCode: string) => string | void | Promise<string | void>;
}

export type UserOptions = Partial<Options>;
