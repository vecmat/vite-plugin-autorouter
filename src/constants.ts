export const MODULE_VIRTUAL_ID = "virtual:autorouter"
export const MODULE_REQUEST_ID = "/@vite-plugin-autorouter/generated-pages";