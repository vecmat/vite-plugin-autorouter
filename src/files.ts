import fg from "fast-glob";
import {  ResolvedOptions } from "./types";
import { extensionsToGlob } from "./utils";

function getIgnore(exclude: string[]) {
    return ["node_modules", ".git", "**/__*__/**", ...exclude];
}

/**
 * Resolves the page dirs for its for its given globs
 */
export function getPageDirs(pageDirOptions: string, root: string, exclude: string[]): string[] {
    const dirs = fg.sync(pageDirOptions, {
        ignore: getIgnore(exclude),
        onlyDirectories: true,
        dot: true,
        unique: true,
        cwd: root,
    });

    return dirs;
}

/**
 * Resolves the files that are valid pages for the given context.
 */
export function getPageFiles(path: string, options: ResolvedOptions): string[] {
    const { exclude, extensions } = options;

    const ext = extensionsToGlob(extensions);

    const files = fg.sync(`**/*.${ext}`, {
        ignore: getIgnore(exclude),
        onlyFiles: true,
        cwd: path,
    });

    return files;
}
