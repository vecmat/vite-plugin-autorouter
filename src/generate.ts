import { parse } from "path";
import { stringifyRoutes } from "./stringify";
import { isDynamicRoute, isCatchAllRoute } from "./utils";
import { Route, ResolvedOptions, ResolvedPages, ResolvedPage } from "./types";

export function generateClientCode(routes: Route[], options: ResolvedOptions) {
    const { imports, stringRoutes } = stringifyRoutes(routes, options);
    return `${imports.join(";\n")};\n\nconst routes = ${stringRoutes};\n\nexport default routes;`;
}


// 排序规则
// layout >  slash > strlen
export function sortPage(pages: ResolvedPages) {
    return (
        [...pages]
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            .map(([_, value]) => value)
            .sort((a, b) => {
                if (a.parents && !b.parents) return 1;
                if (!a.parents && b.parents) return -1;
                let slasha = (a.path.match(/\//g) || []).length;
                let slashb = (b.path.match(/\//g) || []).length;
                if (slasha != slashb) {
                    return slasha - slashb;
                } else {
                    return a.path.length - b.path.length;
                }
            })
    );
}

// match > slash >strlen
export function sortRouter(stack: Route[]) {
    return stack
    .sort((a, b) => {
        let matcha  = a.path.replace(/[^\:\/]/g,"").indexOf(":") 
        matcha=matcha<0?999:matcha;
        let matchb  = b.path.replace(/[^\:\/]/g,"").indexOf(":")
        matchb=matchb<0?999:matchb;
        let slasha = (a.path.match(/\//g) || []).length;
        let slashb = (b.path.match(/\//g) || []).length;
        if(matcha != matchb){
            return matchb - matcha ;
        }else{
            if (slasha != slashb) {
                return slasha - slashb;
            } else {
                return a.path.length - b.path.length;
            }
        }
    })
}


function insertRouter(stack: Route[], parent: string, route: Route):boolean {
    // todo duplicate check ？
    // insert to root
    let inserted = false;
    if (!parent) {
        route.path = "$" + route.path;
        route.chain = "$" + route.chain;
        stack.push(route);
        return true;
    }
    // insert children router
    for(let node of stack){
        // console.log(node.chain,">>>",parent)
        if (node.chain?.endsWith(parent)) {
            inserted = true;
            route.name = node.name +"-"+  route.name;
            route.chain = node.chain +  route.path;
            node.children = node.children || [];
            node.children.push(route);
        }
        if (node.children) {
            if(insertRouter(node.children, parent, route)){
                inserted = true;
            }
        }
    };
    return inserted;
}


function prepareRoutes(stack: Route[], options: ResolvedOptions, root?: boolean) {
    let repeat = new Set();
    for (let node of stack) {
        if (repeat.has(node.path)) {
            throw new Error(`[vite-plugin-auturouter] duplicate route for ${node.component}`);
        }
        repeat.add(node.path);
        delete node.chain;
        if (root) {
            node.path = node.path.replace(/^\$/, "");
        } else {
            node.path = node.path.replace(/^\//, "");
        }
        if (!options.react) {
            node.props = true;
        } else {
            node.routes = node.children;
            delete node.children;
            node.exact = true;
            delete node.name;
        }
        if (node.children) {
            // delete node.name;
            node.children = prepareRoutes(node.children, options);
        }
        Object.assign(node, options.extendRoute?.(node) || {});
    }
    return  sortRouter(stack);
}

export function generateRoutes(pages: ResolvedPages, options: ResolvedOptions): Route[] {
    const { nuxtStyle } = options;

    let stack: Route[] = [];
    const sortpages = sortPage(pages);
    // console.log(sortpages)
    sortpages.forEach((page: ResolvedPage) => {
        let { name, path, parents, component } = page;
        if (typeof parents === "string") {
            parents = [parents];
        }
        parents.map((parent: string) => {
            const route: Route = {
                path: path,
                name: name,
                chain: parent+path,
                component,
            };
            let inserted = insertRouter(stack, parent, route);
            // 深度目录直接创建多层级路由
            // insert for depth router
            if (!inserted) {
                path = path.replace(/^([^/$])/,"/$1");
                parent =  parent.replace(/^\$/,"");
                parent =  parent.replace(/^([^/$])/,"/$1");
                route.path = parent+path
                route.path = "$" +  parent+path,
                stack.push(route);
            }
        });
    });
    // console.log(JSON.stringify(stack, null, 4));
    const finalRoutes = prepareRoutes(stack, options, true);
    // console.log(JSON.stringify(finalRoutes, null, 4));
    return finalRoutes;
}

