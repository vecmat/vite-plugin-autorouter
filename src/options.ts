import { resolve } from "path";
import { UserOptions, ResolvedOptions } from "./types";
import { getPageDirs } from "./files";
import { toArray, slash } from "./utils";

function resolvePageDirs(pagesDir: string[], root: string, exclude: string[]) {
    pagesDir = toArray(pagesDir);
    return pagesDir.flatMap((pagesDir) => {
        pagesDir = slash(resolve(root, pagesDir)).replace(`${root}/`, "");
        return getPageDirs(pagesDir, root, exclude);
    });
}

export function resolveOptions(userOptions: UserOptions, viteRoot?: string): ResolvedOptions {
    const {
        pagesDir = ["src/pages"],
        routeBlockLang = "json5",
        exclude = [],
        syncIndex = true,
        replaceSquareBrackets = false,
        nuxtStyle = false,
        react = false,
        extendRoute,
        onRoutesGenerated,
        onClientGenerated,
    } = userOptions;

    const root = viteRoot || slash(process.cwd());

    const importMode = userOptions.importMode || (react ? "sync" : "async");

    const extensions = userOptions.extensions || (react ? ["tsx", "jsx"] : ["vue", "ts", "js"]);

    const extensionsRE = new RegExp(`\\.(${extensions.join("|")})$`);

    const resolvedPagesDir = resolvePageDirs(pagesDir, root, exclude);

    const resolvedOptions: ResolvedOptions = {
        pagesDir: resolvedPagesDir,
        root,
        react,
        exclude,
        syncIndex,
        nuxtStyle,
        extensions,
        importMode,
        extensionsRE,
        routeBlockLang,
        replaceSquareBrackets,
        extendRoute,
        onRoutesGenerated,
        onClientGenerated,
    };

    return resolvedOptions;
}
