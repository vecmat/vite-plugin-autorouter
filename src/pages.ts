import { getPageFiles } from "./files";
import { join, extname, resolve ,basename,dirname} from "path";
import {  ResolvedOptions, ResolvedPages, ResolvedPage } from "./types";
import { getRouteBlock, routeBlockCache, toArray, slash } from "./utils";


async function setPage(pages: ResolvedPages, dir:string, file: string, options: ResolvedOptions) {
    let extension = extname(file).slice(1);
    let parents = dirname(file).replace(dir,"")
    let filepath = slash(resolve(options.root, file));
    let filename = basename(file).replace(options.extensionsRE, "");
    filename = filename.replace(/(\[([^\[\]]+)\])/g,":$2")
    parents =  parents.replace(/(\[([^\[\]]+)\])/g,":$2")
    filename = filename.replace(/(\[(\.\.\.)\])/g,":any(.*)*")
    parents =  parents.replace(/(\[(\.\.\.)\])/g,":any(.*)*")
    let page : ResolvedPage = { 
        name:`${filename}` ,
        path:`/${filename}` , 
        parents : `${parents}`,
        component : `/${file}`,
     } ;
    let block : Record<string, any> =  {};
    if(["vue", "md"].includes(extension)){
        block = await getRouteBlock(filepath, options)
    }
    Object.assign(page,block)
    page.path =   page.path.replace(/^([^/$])/,"/$1");
    pages.set(file,page);
}

export async function resolvePages(options: ResolvedOptions) {
    const dirs = toArray(options.pagesDir);
    const pages = new Map<string, ResolvedPage>();
    const pageDirFiles = dirs.map((dir) => {
        const pagePath = slash(resolve(options.root, dir));
        return {
            dir:dir,
            files: getPageFiles(pagePath, options),
        };
    });
    for (const row of pageDirFiles) {
        for (const file of row.files) {
            await setPage(pages,row.dir, join(row.dir , file), options);
        }
    }
    return pages;
}

export async function addPage(pages: ResolvedPages, file: string, options: ResolvedOptions) {
    file = file.replace(options.root, "");
    const pageDir = options.pagesDir.find((i) => {
        file.startsWith(`/${i}`)
    });
    if (!pageDir) return;
    await setPage(pages,pageDir , file ,options);
}

export async function updatePage(pages: ResolvedPages, file: string , options: ResolvedOptions ) {
    const page = pages.get(file);
    if (page) {
        const customBlock = routeBlockCache.get(file) || null;
        page.customBlock = customBlock;
        pages.set(file, page);
        console.log(options)
    }
}

export function removePage(pages: ResolvedPages, file: string) {
    pages.delete(file);
}

