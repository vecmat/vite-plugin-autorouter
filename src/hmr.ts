import { ViteDevServer } from "vite";
import { getPagesVirtualModule, isRouteBlockChanged, isTarget, debug, slash } from "./utils";
import { removePage, addPage, updatePage } from "./pages";
import { ResolvedOptions, ResolvedPages } from "./types";

export function handleHMR(server: ViteDevServer, pages: ResolvedPages, options: ResolvedOptions, clearRoutes: () => void) {
    const { ws, watcher, } = server;

    function fullReload() {
        // invalidate module
        getPagesVirtualModule(server);
        clearRoutes();
        ws.send({
            type: "full-reload",
        });
    }

    watcher.on("add", async(path) => {
        if(/(vue.md)$/.test(path)) return
        const file =  path.replace(options.root, "");
        if (isTarget(file, options)) {
            await addPage(pages, file, options);
            debug.hmr("add", file);
            fullReload();
        }
    });
    watcher.on("unlink", (path) => {
        if(/(vue.md)$/.test(path)) return
        const file =  path.replace(options.root, "");
        if (isTarget(file, options)) {
            removePage(pages, file);
            debug.hmr("remove", file);
            fullReload();
        }
    });
    watcher.on("change", async(path) => {
        if(/(vue.md)$/.test(path)) return
        const file =  path.replace(options.root, "");
        if (isTarget(file, options) && !options.react) {
            const needReload = await isRouteBlockChanged(file, options);
            if (needReload) {
                updatePage(pages, file, options);
                debug.hmr("change", file);
                fullReload();
            }
        }
    });
}
