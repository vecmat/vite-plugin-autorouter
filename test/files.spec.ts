import { join } from 'path'
import { slash } from '../src/utils'
import { resolveOptions } from '../src/options'
import { getPageFiles, getPageDirs } from '../src/files'

const options = resolveOptions({})
const testPagesDir = 'test/assets/pages'
const testDeepPagesDir = 'test/assets/deep-pages'

describe('Get files', () => {
  test('Pages file', async() => {
    const files = getPageFiles(testPagesDir, options)
    expect(files.sort()).toMatchSnapshot('page files')
  })
})

describe('Get page dirs', () => {
  test('With glob', async() => {
    const dir = slash(join(testDeepPagesDir, '**', 'pages'))
    const dirs = getPageDirs(dir, options.root, options.exclude)
    expect(dirs.sort()).toMatchSnapshot('glob dirs')
  })
})
