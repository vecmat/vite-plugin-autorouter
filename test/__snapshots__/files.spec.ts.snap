// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get files Pages file: page files 1`] = `
Array [
  "404.vue",
  "about.vue",
  "about/index.vue",
  "about/test.vue",
  "blog/[id].vue",
  "blog/[id]/index.vue",
  "blog/index.vue",
  "blog/liar.vue",
  "blog/today.vue",
  "blog/today/index.vue",
  "catch/aaa/_.vue",
  "catch/bbb/_userId.vue",
  "catch/ccc/[...].vue",
  "catch/ddd/[...all].vue",
  "catch/eee/[userId].vue",
  "catch/fff/pre[userId].vue",
  "catch/index.vue",
  "components.vue",
  "index.vue",
  "sensor.vue",
  "sensor/404.vue",
  "sensor/index.vue",
]
`;

exports[`Get page dirs With glob: glob dirs 1`] = `
Array [
  "test/assets/deep/bar/pages",
  "test/assets/deep/foo/pages",
]
`;
